### PR TITLE
Sorting algorithms visualised

### PR DESCRIPTION
**Pull request by callumacrae**

### CODE REVIEW

File: articles/sorting-algorithms.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
+# Sorting algorithms visualised
+- 19th December 2016
+
+Sometimes you or a computer needs to sort a list of items—say, a list of names into alphabetical order, or a list of objects into order of length. To sort this list, especially if it is a large list, you need some kind of process: for example, if you were sorting a list of names into alphabetic order, you could put all the names beginning with A together, then all the Bs together, and then you could sort them by the second letter: all the AAs together, then all the ABs together, until they're sorted.
```
Comment (ChrisDBrown) [inline articles/sorting-algorithms.md]: "items - say a list of names..."
"some kind of process - for example, if..."
Split last line in two - "Then all the Bs together. Then you'd sort them by the second letter - all the AAs, ABs, ACs, until all names are sorted."

File: articles/sorting-algorithms.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+Algorithms used to sort a list of items are called sorting algorithms, and in this article I'm going to explain and visualise a few of the most well known sorting algorithms. Each section will contain one sorting algorithm, a description of the algorithm, and then a visualisation of the algorithm, sorting different length bars (seen below) from shortest to longest, powered by [D3.js]. There are some controls to the bottom right of the screen: it might help if you adjust the speed of the visualisation a bit.
```
Comment (ChrisDBrown) [inline articles/sorting-algorithms.md]: "visualisation of the algorithm sorting different length bars from shortest to longest, as seen below. These are powered by [D3.js] and can be adjusted with the controls to the bottom right of the screen."

File: articles/sorting-algorithms.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+The orange bars are the bars being compared, and the darker blue bars are the bars that have finished sorting. Watch to the see the largest unsorted bar being grabbed and "bubbling up" the list:
+
+<svg class="sorting-demo" id="bubble" height="300" width="800" viewBox="0 0 800 300"></svg>
+
+----
+
+## Selection sort
+
+Selection sort works by iterating through the list, finding the smallest unsorted item. Once it has found it, it swaps the item with the left-most unsorted item, marks it as sorted, and starts the search again.
```
Comment (ChrisDBrown) [inline articles/sorting-algorithms.md]: "through the list and finding the..." (switch comma with and)
"found it it swaps the..." (remove comma)

File: articles/sorting-algorithms.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+<svg class="sorting-demo" id="merge" height="300" width="800" viewBox="0 0 800 300"></svg>
+
+An algorithm which divides the problem into smaller problems and solves them recursively is known as a divide and conquer algorithm.
+
+----
+
+## Shell sort
+
+Shell sort is a variation on insertion sort, which was explained above. It's a different in that it runs through the list multiple times, and starts by comparing elements that are far apart from each other—the distance between them being known as the "gap".
```
Comment (ChrisDBrown) [inline articles/sorting-algorithms.md]: Worth re-ordering to put this directly after insertion sort

File: articles/sorting-algorithms.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+The dark blue elements are elements that have been marked as sorted, the green element is the pivot element, and the orange elements are the high and low elements.
+
+<svg class="sorting-demo" id="quick" height="300" width="800" viewBox="0 0 800 300"></svg>
+
+----
+
+## Heap sort
+
+Heap sort is probably the most complicated of all the algorithms explained in this article, as it relies on some other concepts from Computer Science—heaps.
```
Comment (ChrisDBrown) [inline articles/sorting-algorithms.md]: "Science - heaps."

File: articles/sorting-algorithms.md [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+The algorithm works in two stages: in the first stage, it takes unprocessed items one at a time, and adds them one at the time to a list representation of a heap—a tree structure where the parent nodes are always bigger than the child nodes. [This video on heaps][xoax heaps] is a good watch, if you're unfamiliar with the concept.
+
+In the second stage, once the list has become a heap, the largest element—the first element in the list, now—is repeatedly swapped with the last unsorted item in the list. It is then marked as completed, and the item that was just swapped onto the top of the heap repeatedly swapped with its smallest child until the heap is valid again.
```
Comment (ChrisDBrown) [inline articles/sorting-algorithms.md]: "largest element - now the first element in the list - is repeatedly...."

File: articles/sorting-algorithms.md (hunk 5 above)
Comment (callumacrae) [inline articles/sorting-algorithms.md]: it's an em dash, github just doesn't display it like that! it's wide and designed to be used like that—like this.

Comment (ChrisDBrown) [inline articles/sorting-algorithms.md]: lol, thanks github.

**Pull request merged**